### PR TITLE
Fix state-db pinning

### PR DESCRIPTION
**Pull request by arkpar**
A follow up to #12902. Previous PR did not account for the possibility of multiple blocks finalized in a single transaction. This version accumulates all canonicalized changes pinned in memory, until explicitly released with the `sync` function.  Similar to `apply_changes` function that existed before the state-db refactoring.

### CODE REVIEW

**Review by cheme — APPROVED** [review]

**Review by lexnv — APPROVED** [review]

File: client/state-db/src/lib.rs [hunk 1]
```diff
@@ -573,6 +577,12 @@ impl<BlockHash: Hash, Key: Hash, D: MetaDb> StateDb<BlockHash, Key, D> {
 		self.db.write().unpin(hash)
 	}
 
+	/// Confirm that all changes made to commit sets are on disk. Allow for temporarily pinned
+	/// blocks to be released.
+	pub fn sync(&self) {
```
[the remark this answers is not in the document]
Comment (arkpar) [reply client/state-db/src/lib.rs:582]: I went for a more general name. It basically means that the in-memory state is synchronized with the on-disk state.

File: client/state-db/src/noncanonical.rs [hunk 2]
```diff
@@ -350,6 +350,16 @@ impl<BlockHash: Hash, Key: Hash> NonCanonicalOverlay<BlockHash, Key> {
 		self.last_canonicalized.as_ref().map(|&(_, n)| n)
 	}
 
+	pub fn sync(&mut self) {
+		let mut pinned = std::mem::take(&mut self.pinned_canonincalized);
+		for hash in pinned.iter() {
+			self.unpin(hash)
+		}
+		pinned.clear();
```
Comment (bkchr) [inline client/state-db/src/noncanonical.rs:358]:
```suggestion
		self.pinned_canonincalized.drain(..).for_each(|h| self.unpin(h));
```

**Review by bkchr — APPROVED** [review]

File: client/state-db/src/noncanonical.rs [hunk 3]
```diff
@@ -350,6 +350,16 @@ impl<BlockHash: Hash, Key: Hash> NonCanonicalOverlay<BlockHash, Key> {
 		self.last_canonicalized.as_ref().map(|&(_, n)| n)
 	}
 
+	pub fn sync(&mut self) {
```
Comment (bkchr) [inline client/state-db/src/noncanonical.rs:353]: Would also be better with some docs.

File: client/state-db/src/noncanonical.rs (hunk 2 above)
Comment (arkpar) [reply client/state-db/src/noncanonical.rs:358]: This won't work because `drain` borrows `self` and `unpin` requires `&mut self`

**Comment by arkpar** [comment]
bot merge

File: client/state-db/src/noncanonical.rs (hunk 2 above)
Comment (bkchr) [reply client/state-db/src/noncanonical.rs:360]: Good point, but the call to `clear` is also not really required.

**Pull request merged**